presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java

```diff
@@ -87,75 +63,146 @@ public CassandraClusteringPredicatesExtractor(List<CassandraColumnHandle> cluste
+                                            + CassandraCqlUtils.cqlValue(toCQLCompatibleString(range.getHigh().getValue()),
+                                            columnHandle.getCassandraType()));
+                                    break;
+                                default:
+                                    throw new AssertionError("Unhandled bound: " + range.getHigh().getBound());
+                            }
                         }
-                        return ImmutableSet.of();
-                    },
-                    allOrNone -> ImmutableSet.of());
+                    }
+                }
 
-            if (!values.isEmpty()) {
-                clusteringColumnValues.add(values);
+                if (singleValues.size() == 1) {
+                    predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " = " + singleValues.get(0);
+                }
+                else if (singleValues.size() > 1) {
+                    if (!isPredicateSupported(cassandraVersion)) {
+                        return "Not Supported";
+                    }
+                    predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " IN ("
+                            + Joiner.on(",").join(singleValues) + ")";
+                }
+                else if (!rangeConjuncts.isEmpty()) {
+                    predicate = Joiner.on(" AND ").join(rangeConjuncts);
+                }
+                return predicate;
+            }, discreteValues -> {
+                if (discreteValues.isWhiteList()) {
+                    ImmutableList.Builder<Object> discreteValuesList = ImmutableList.builder();
+                    if (!isPredicateSupported(cassandraVersion)) {
+                        return "Not Supported";
+                    }
+                    for (Object discreteValue : discreteValues.getValues()) {
+                        discreteValuesList.add(CassandraCqlUtils.cqlValue(toCQLCompatibleString(discreteValue),
+                                columnHandle.getCassandraType()));
+                    }
+                    String predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " IN ("
+                            + Joiner.on(",").join(discreteValuesList.build()) + ")";
+                    return predicate;
+                }
+                return null;
+            }, allOrNone -> null);
+
+            if (predicateString != null) {
+                // Check for Cassandra version older than 2.2.x
+                if (predicateString.equals("Not Supported")) {
+                    break;
+                }
+                clusteringColumnSql.add(predicateString);
                 domainsBuilder.put(columnHandle, domain);
+                // Check for last clustering column should be only restricted by range condition
+                if (predicateString.contains(">") || predicateString.contains("<")) {
```

Comment:
The restriction that the only last cluster key column should have the IN predicate is true only for Cassandra versions < 2.2. So for Cassandra versions > 2.2, this should check shouldn't be there.